Super Cache: Fixed delete cache button trigger hook from admin panel

Dalias96 opened the pull request:
After that change hook wpsc_after_delete_cache_admin_bar will be fired also when pressing button from admin panel

<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes Automattic/wp-super-cache#999

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
*

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No, it doesn't.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Add any action to hook `wpsc_after_delete_cache_admin_bar`
* Delete cache via button from front page
* Delete cache via buttom from admin panel
* Hook should be triggered in both scenarios

Dalias96 committed:
Fixed delete cache button trigger hook from admin panel

After that change hook wpsc_after_delete_cache_admin_bar will be fired also when pressing button from admin panel

donnchawp commented on projects/plugins/super-cache/inc/delete-cache-button.php:
```diff
@@ -86,6 +86,17 @@ function wpsc_admin_bar_delete_cache_ajax() {
 		if ( defined( 'WPSCDELETEERROR' ) ) {
 			return json_decode( constant( 'WPSCDELETEERROR' ) );
 		} else {
+			$req_path = !empty($_POST['path']) ? $_POST['path'] : '/';
```
Can you add spaces in the empty() call here, so it follows the WordPress coding conventions?  It should look like this:
`! empty( $_POST['path'] )`

It's probably worth sanitizing as well, since that's what happens when `wpsc_after_delete_cache_admin_bar` is used further down in the code.  the default path is '' in the other action, so we should probably change that too. 
The line will look like this:
`$req_path = ! empty( $_POST['path'] ) ? sanitize_text_field( stripslashes( $_POST['path'] ) ) : '';`

Dalias96 replied:
I just copied full line from code below, to keep consistency.

donnchawp commented on projects/plugins/super-cache/inc/delete-cache-button.php:
```diff
@@ -86,6 +86,17 @@ function wpsc_admin_bar_delete_cache_ajax() {
 			return json_decode( constant( 'WPSCDELETEERROR' ) );
 		} else {
+			$req_path = !empty($_POST['path']) ? $_POST['path'] : '/';
+			$referer = !empty($_SERVER['HTTP_REFERER']) ? $_SERVER['HTTP_REFERER'] : admin_url();
```
It's probably safer to use `wp_get_referer()` rather than accessing $_SERVER directly.

Dalias96 replied:
good point, changed that :)

donnchawp reviewed (changes requested):
Thank you so much for your PR to add this missing code to the delete button. There are a few small things that need to be changed and then we can get this merged.

Dalias96 committed:
Changed req_path and referer values

Changed req_path and referer values in order to keep consistency with WP coding standards and keep sanitization.

Dalias96 commented on projects/plugins/super-cache/inc/delete-cache-button.php:
```diff
@@ -86,6 +86,17 @@ function wpsc_admin_bar_delete_cache_ajax() {
+			/**
+			 * Hook into the cache deletion process after a successful cache deletion from the admin bar button.
+			 *
+			 * @since 1.12
```
@donnchawp maybe different version should be here?

donnchawp replied:
Yeah, can you change that version to this: `$$next-version$$`

Dalias96 commented:
@donnchawp any update on the above?

donnchawp commented on projects/plugins/super-cache/inc/delete-cache-button.php:
```diff
@@ -86,6 +86,17 @@ function wpsc_admin_bar_delete_cache_ajax() {
 		if ( defined( 'WPSCDELETEERROR' ) ) {
 			return json_decode( constant( 'WPSCDELETEERROR' ) );
 		} else {
+			$req_path = isset( $_POST['path'] ) ? sanitize_text_field( stripslashes( $_POST['path'] ) ) : '';
```
PHP Code Sniffer found some (longstanding) issues here that we should look at. The first is the nonce check. Can you add this code right at the top of the function?

```
	if ( ! isset( $_POST['path'] ) ) {
		$_POST['path'] = '/';
	}

	if ( ! isset( $_POST['nonce'] ) ) {
		wp_send_json_error( 'missing nonce.', 401 );
	}

	if ( ! wp_verify_nonce( $_POST['nonce'], 'delete-cache-' . wp_unslash( $_POST['path'] ) . '_1' ) ) { // "1" for admin
		wp_send_json_error( 'Invalid nonce.', 401 );
	}
```
That takes care of the nonce check.

Can you change `stripslashes` on line 89 to `wp_unslash`? That will satisfy the other PHPCS check.

donnchawp replied:
One more change to support returning an error. The file delete-cache-button.js needs the following code around line 24, just before the "complete:" section
```
error: function ( msg ) {
    console.log( 'Error deleting cache' );
},
```

Make sure it aligns with the rest of the code, please.

Sorry about making this patch more complicated. Jetpack has strict code checks, and they notice things that are missing.

donnchawp commented on projects/plugins/super-cache/inc/delete-cache-button.php:
```diff
@@ -86,6 +86,17 @@ function wpsc_admin_bar_delete_cache_ajax() {
 			return json_decode( constant( 'WPSCDELETEERROR' ) );
 		} else {
+			$req_path = isset( $_POST['path'] ) ? sanitize_text_field( stripslashes( $_POST['path'] ) ) : '';
+			$referer = wp_get_referer();
```
PHPCS wants "=" signs to line up on consecutive lines. Can you line them up on lines 89 and 90, please?
